app/Foliage/CmdBuild.hs

```diff
@@ -113,9 +113,22 @@ buildAction
             cf <- prepareIndexPkgCabal pkgId cabalFileTimestamp originalCabalFilePath
 
             -- all revised cabal files, with their timestamp
-            revcf <- for cabalFileRevisions $ uncurry (prepareIndexPkgCabal pkgId)
-
-            return $ cf : revcf
+            revcf <-
+              foldM
+                ( \(lastRevisionPath, cabalFiles) (timestamp, revisionPath) -> do
+                    if takeExtension originalCabalFilePath `elem` [".diff", ".patch"]
+                      then do
+                        patchedRevisionPath <- applyPatch lastRevisionPath revisionPath
+                        preparedCabal <- prepareIndexPkgCabal pkgId timestamp patchedRevisionPath
```

Comment:
I think this will work weirdly. `prepareIndexPkgCabal` calls `need` on the cabal file path. That's not going to work properly when it's a temporary patched file: the patched file won't exist in future, and it won't track the dependencies properly.

I think probably we want to do this inside `prepareIndexPkgCabal`: give it the base cabal file and a list of patch files, and then do the patching after `need`-ing the base file and all the patch files.